### PR TITLE
REF: unstack

### PR DESCRIPTION
**Pull request by jbrockmendel**
cc @jreback this moves a small amount of the unstack logic up the call stack, which is worthwhile, but the big wins available are if you can help me figure out the questions in inline comments.

### CODE REVIEW

File: pandas/core/reshape/reshape.py [hunk 1]
```diff
@@ -205,6 +205,9 @@ def get_new_values(self, values, fill_value=None):
 
         # we can simply reshape if we don't have a mask
         if mask_all and len(values):
+            # TODO: Under what circumstances can we rely on sorted_values
```
Comment (jreback) [inline pandas/core/reshape/reshape.py:208]: I think if the we are sorted *already* then you can simplify, but I am not sure we can make that guarantee.

**Comment by jreback** [comment]
looks fine. you can try experimenting if you try to unstack on a non-sorted index and raising and see what breaks.

File: pandas/core/internals/blocks.py [hunk 2]
```diff
@@ -1403,17 +1401,17 @@ def _unstack(self, unstacker, new_columns, fill_value, value_columns):
...
         )
 
         mask = mask.any(0)
+        # TODO: in all tests we have mask.all(); can we rely on that?
+
         new_values = new_values.T[mask]
         new_placement = new_placement[mask]
 
-        blocks = [make_block(new_values, placement=new_placement)]
+        blocks = [self.make_block_same_class(new_values, placement=new_placement)]
```
Comment (simonjayhawkins) [inline pandas/core/internals/blocks.py:1414]: @jbrockmendel I think it's just this line that caused the regression in #37115 and that we don't need to revert the complete PR.

Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:1414]: what kind of block(s) do you end up with instead?

Comment (simonjayhawkins) [reply pandas/core/internals/blocks.py:1414]: corrupted? an IntBlock with dtype float64

```
>>> pd.__version__
'1.2.0.dev0+1432.g5fdf642368'
>>>
>>> df1 = pd.DataFrame(
...     {
...         "a": ["A", "A", "B"],
...         "b": ["ca", "cb", "cb"],
...         "v": [10] * 3,
...     }
... )
>>> df1 = df1.set_index(["a", "b"])
>>> df1["is_"] = 1
>>> df1
       v  is_
a b
A ca  10    1
  cb  10    1
B cb  10    1
>>>
>>> df1._data
BlockManager
Items: Index(['v', 'is_'], dtype='object')
Axis 1: MultiIndex([('A', 'ca'),
            ('A', 'cb'),
            ('B', 'cb')],
           names=['a', 'b'])
IntBlock: slice(0, 1, 1), 1 x 3, dtype: int64
IntBlock: slice(1, 2, 1), 1 x 3, dtype: int64
>>>
>>> df2 = df1.unstack("b")
>>> df2
      v        is_
b    ca    cb   ca   cb
a
A  10.0  10.0  1.0  1.0
B   NaN  10.0  NaN  1.0
>>>
>>> df2._data
BlockManager
Items: MultiIndex([(  'v', 'ca'),
            (  'v', 'cb'),
            ('is_', 'ca'),
            ('is_', 'cb')],
           names=[None, 'b'])
Axis 1: Index(['A', 'B'], dtype='object', name='a')
IntBlock: slice(0, 2, 1), 2 x 2, dtype: float64
IntBlock: slice(2, 4, 1), 2 x 2, dtype: float64
>>>
```

Comment (simonjayhawkins) [reply pandas/core/internals/blocks.py:1414]: run the test suite with just this change reverted and all tests pass. will put up a PR shortly.

**Pull request merged**